contracts/src/registry/PermissionedRegistry.sol

```diff
@@ -2,55 +2,295 @@
         external
-        onlyTokenOwner(tokenId)
-        withSubregistryFlags(tokenId, FLAG_SUBREGISTRY_LOCKED, 0)
+        override
+        onlyNonExpiredTokenRoles(tokenId, ROLE_SET_SUBREGISTRY)
     {
-        (, uint96 _flags) = datastore.getSubregistry(tokenId);
-        datastore.setSubregistry(tokenId, address(registry), _flags);
+        (, uint64 expires, uint32 tokenIdVersion) = datastore.getSubregistry(tokenId);
+        datastore.setSubregistry(tokenId, address(registry), expires, tokenIdVersion);
     }
 
     function setResolver(uint256 tokenId, address resolver)
         external
-        onlyTokenOwner(tokenId)
-        withSubregistryFlags(tokenId, FLAG_RESOLVER_LOCKED, 0)
+        override
+        onlyNonExpiredTokenRoles(tokenId, ROLE_SET_RESOLVER)
     {
-        (, uint96 _flags) = datastore.getResolver(tokenId);
-        datastore.setResolver(tokenId, resolver, _flags);
+        datastore.setResolver(tokenId, resolver, 0, 0);
+    }
+
+    function getNameData(string calldata label) external view returns (uint256 tokenId, uint64 expiry, uint32 tokenIdVersion) {
+        uint256 canonicalId = NameUtils.labelToCanonicalId(label);
+        (, expiry, tokenIdVersion) = datastore.getSubregistry(canonicalId);
+        tokenId = _constructTokenId(canonicalId, tokenIdVersion);
+    }
+
+    function getExpiry(uint256 tokenId) external view override returns (uint64) {
+        (, uint64 expires, ) = datastore.getSubregistry(tokenId);
+        return expires;
+    }
+
+    function supportsInterface(bytes4 interfaceId) public view virtual override(BaseRegistry, EnhancedAccessControl, IERC165) returns (bool) {
+        return interfaceId == type(IPermissionedRegistry).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    function getTokenIdResource(uint256 tokenId) public pure override returns (bytes32) {
+        return bytes32(NameUtils.getCanonicalId(tokenId));
     }
 
-    function flags(uint256 tokenId) external view returns(uint96) {
-        (, uint96 _flags) = datastore.getSubregistry(tokenId);
-        return _flags;
+    function getResourceTokenId(bytes32 resource) public view override returns (uint256) {
+        uint256 canonicalId = uint256(resource);
+        (, , uint32 tokenIdVersion) = datastore.getSubregistry(canonicalId);
+        return _constructTokenId(canonicalId, tokenIdVersion);
+    }
+
+    // Internal/private methods
+
+    /**
+     * @dev Override the base registry _update function to transfer the roles to the new owner when the token is transferred.
+     */
+    function _update(address from, address to, uint256[] memory ids, uint256[] memory values) internal virtual override {
+        super._update(from, to, ids, values);
+
+        for (uint256 i = 0; i < ids.length; ++i) {
+            /*
+            in _regenerateToken, we burn the token and then mint a new one. This flow below ensures the roles go from owner => zeroAddr => owner
```

Comment:
It's not clear to me what this comment means. Isn't the order the reverse of what's implied?

Reply:
the flow is:

1. burn
2. mint

in (1) the roles are copied from owner to zero address, in (2) they are copied from zero address back to the owner.